reporter_test.go

```diff
@@ -465,9 +496,115 @@ func TestDiffFile(t *testing.T) {
 				t.Errorf("diffFile() error: %v", err)
 			default:
 				if gotDiff != tc.wantDiff {
-					t.Errorf("diffFile() diff: got=%s, want=%s", gotDiff, tc.wantDiff)
+					t.Errorf("diffFile() diff: got=%q, want=%q", gotDiff, tc.wantDiff)
 				}
 			}
 		})
 	}
 }
+
+func TestCompare(t *testing.T) {
+	testCases := []struct {
+		desc   string
+		before *fspb.Walk
+		after  *fspb.Walk
+		handle func(*Report, error) bool
+	}{
+		{
+			desc:   "nil before",
+			before: nil,
+			after: &fspb.Walk{
+				File: []*fspb.File{
+					&fspb.File{Path: "/a/b/c", Info: &fspb.FileInfo{}},
+				},
+			},
+			handle: func(r *Report, _ error) bool {
+				return len(r.Added) == 1
+			},
+		}, {
+			desc: "empty after",
+			before: &fspb.Walk{
+				Id: "1",
+				File: []*fspb.File{
+					&fspb.File{Path: "/a/b/c", Info: &fspb.FileInfo{}},
+				},
+			},
+			after: &fspb.Walk{Id: "2"},
+			handle: func(r *Report, _ error) bool {
+				return len(r.Deleted) == 1
+			},
+		}, {
+			desc:   "nil before and after",
+			before: nil,
+			after:  nil,
+			handle: func(_ *Report, err error) bool {
+				return err != nil
+			},
+		}, {
+			desc: "diffs",
+			before: &fspb.Walk{
+				Id: "1",
+				File: []*fspb.File{
+					&fspb.File{Path: "/a/b/c", Info: &fspb.FileInfo{}},
+					&fspb.File{Path: "/e/f/g", Info: &fspb.FileInfo{Size: 4}},
+					&fspb.File{Path: "/x/y/z", Info: &fspb.FileInfo{}},
+				},
+			},
+			after: &fspb.Walk{
+				Id: "2",
+				File: []*fspb.File{
+					&fspb.File{Path: "/b/c/d", Info: &fspb.FileInfo{}},
+					&fspb.File{Path: "/e/f/g", Info: &fspb.FileInfo{Size: 7}},
+					&fspb.File{Path: "/x/y/z", Info: &fspb.FileInfo{}},
+				},
+			},
+			handle: func(r *Report, _ error) bool {
+				return len(r.Deleted) == 1 && len(r.Added) == 1 && len(r.Modified) == 1
+			},
+		}, {
+			desc: "ignore",
+			before: &fspb.Walk{
+				Id: "1",
+				File: []*fspb.File{
+					&fspb.File{Path: "/ignore/a", Info: &fspb.FileInfo{}},
+				},
+			},
+			after: &fspb.Walk{
+				Id: "2",
+				File: []*fspb.File{
+					&fspb.File{Path: "/ignore/b", Info: &fspb.FileInfo{}},
+				},
+			},
+			handle: func(r *Report, _ error) bool {
+				return len(r.Deleted) == 0 && len(r.Added) == 0 && len(r.Modified) == 0
+			},
+		}, {
+			desc: "same dir with and without trailing /",
+			before: &fspb.Walk{
+				Id: "1",
+				File: []*fspb.File{
+					&fspb.File{Path: "/a/b/c/", Info: &fspb.FileInfo{IsDir: true}},
+				},
+			},
+			after: &fspb.Walk{
+				Id: "2",
+				File: []*fspb.File{
+					&fspb.File{Path: "/a/b/c", Info: &fspb.FileInfo{IsDir: true}},
+				},
+			},
+			handle: func(r *Report, _ error) bool {
+				return len(r.Deleted) == 0 && len(r.Added) == 0 && len(r.Modified) == 0
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			r := &Reporter{config: &fspb.ReportConfig{ExcludePfx: []string{"/ignore/"}}}
+			report, err := r.Compare(tc.before, tc.after)
+			ok := tc.handle(report, err)
+			if !ok {
+				t.Error("Compare() unexpected result")
```

Comment:
The failure message could be more useful. Reasoning described here in details:
https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

Yoo could add num of expected deletions, additions and modified files to the test cases and then check them here. For example:

```go
if n := len(report.Deleted); n != tc.deleted {
  t.Errorf("len(report.Deleted) = %d; want %d", n, tc.deleted)
}
```
... and same with `report.Added` and `report.Modified`.

Alternatively, you could make the `handle` func field return `error` instead of `bool`, although I'd rename it to something like `verify`:

```go
verify: func(r *Report) error {
  if n := len(r.Deleted); n != 0 {
    return fmt.Errorf("len(r.Deleted) = %d; want 0", n)
  }
  // ... r.Added and r.Modified
},
```

but this version feels more verbose - you'll need to `if n := len(...)` in every `verify` func field - and it looses on usefulness of test failures because reading an error message like `r.Deleted = 1; want 0` make it less clear what the `r` compared to the former.

None of the above would cover cases where you expect `r.Compare(...)` to return an error but maybe those could be factored out to a separate test like `TestCompareErrors`?